TrafficJunction Environment

rafaelmp2 opened the pull request:
Hi, I have implemented the traffic junction environment following the guidelines that you left in the code files.
Please feel free to merge it to your branch if you are interested.

koulanurag commented:
Hi @rafaelmp2 , 
Thanks for your work.  I will take a day or two to review the code, before merging it.

rafaelmp2 commented:
Thanks @koulanurag . Just to clarify this, I realized I left by mistake 2 definitions of `__create_grid()` function. The right one is the second in line 265 (the one you had as a guideline).

koulanurag commented:
You can still go ahead and make commits to the env to remove any duplicates.

I will check it out tomorrow evening.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -44,20 +47,47 @@ class TrafficJunction(gym.Env):
+        mask_size = np.prod(self._agent_view_mask)
+
+        # entry gates where the cars spawn
+        self._entry_gates = [(self._grid_shape[0] // 2, 0), (self._grid_shape[0] - 1, self._grid_shape[1] // 2),\
```
The order of cars should be either clockwise or anti-clockwise.

The correct order would be : [(7, 0), (13, 7), (6, 13), (0, 6)]

koulanurag replied:
Also, I guess, changing the 'self._entry_gates' would require changes in `self._destination`

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -74,12 +104,106 @@ def __init__(self, grid_shape=(14, 14), n_agents=10, n_max=10, arrive_prob=0.5,
         self.observation_space = MultiAgentObservationSpace([spaces.Box(self._obs_low, self._obs_high) for _ in range(self.n_agents)])
-        self.seed()
+
+        self.agents_colors = [(random.randrange(256), random.randrange(256), random.randrange(256)) for _ in range(self.n_agents)]
```
The colors have to be fixed and pre-defined.

This is required so that we have the same colors across multiple instances of the game and even if the random seed is changed.

koulanurag replied:
Let's have a fixed set of colors for all 'n_max' agents.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -44,20 +47,47 @@ class TrafficJunction(gym.Env):
+    def __init__(self, grid_shape=(14, 14), step_cost=0, n_max=4, rcoll=-10, arrive_prob=0.5, full_observable=False):
         self._grid_shape = grid_shape
-        self.n_agents = n_agents
+        self.n_agents = n_max
```
n_agents will be a constant in the case rather than being "n_max" since we haven't defined rest for the code for more than 4 agents.

So, let's keep `self.n_agents=4`. Also, make it private and accessible through a property decorator so that it cannot be changed from outside.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -44,20 +47,47 @@ class TrafficJunction(gym.Env):
 
-        self._entry_gates = [(7, 0), (13, 7), (0, 6), (6, 13)]
+        self._agent_view_mask = (3, 3)
+        mask_size = np.prod(self._agent_view_mask)
```
remove `mask_size` as it's not used anywhere.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -74,12 +104,106 @@ def __init__(self, grid_shape=(14, 14), n_agents=10, n_max=10, arrive_prob=0.5,
+        self._full_obs[self.agent_pos[agent_i][0]][self.agent_pos[agent_i][1]] = PRE_IDS['agent'] + str(agent_i + 1)
+
+
+    def __check_colision(self, pos):
```
spelling  correction  `collision`

koulanurag reviewed (changes requested):
Hi @rafaelmp2,  I have some minor changes in the form of comments. Please, let me know if you would be making these changes.  

This would also require a re-review since many of these suggested changes may lead to downstream changes in your code.

rafaelmp2 commented:
Thanks for the comments @koulanurag , I am willing to address them. I just have a question since you pointed that out regarding the `n_max` and the `n_agents`. When I implemented it I had doubts regarding the number of agents in the environment, so I did it the way it was more convenient to me, i.e., I did it in a way that I could increase the number of agents to more than 4. 

This was commented in the code `'new cars enter the grid with probability p_arrive from each of the four directions. However, the total number of cars at any given time is limited to Nmax = 10'` . So this means that the number of agents is always 4 and the ones that arrive at probability `p_arrive` are not agent-controlled? They are just randomly moving "obstacles"? In my code I consider all of them agent-controlled cars (the environment can have that many agents). If you could please clarify this then I will address it according to your ideas.

koulanurag commented:
You are right.
It's my bad! I totally forgot about the notion of adding more cars into the environment.

We can keep 'n_max' as an attribute.

rafaelmp2 commented:
Thanks, that is clear to me now. I have addressed the comments you left here. If you feel anything else needs to be changed please let me know.

Thanks again for your feedback.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -31,6 +32,7 @@ class TrafficJunction(gym.Env):
     the simulation in any other way. To discourage a traffic jam, each car gets reward of `τrtime = −0.01τ`
     at every time step, where `τ` is the number time steps passed since the car arrived. Therefore, the total
     reward at time t is
+    # Todo: complete the reward information
```
I will update this information.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -73,13 +116,96 @@ def __init__(self, grid_shape=(14, 14), n_agents=10, n_max=10, arrive_prob=0.5,
+        self._full_obs = self.__create_grid()
+
+        for agent_i in range(self.n_agents):
+            while True:
```
In general, try to avoid infinite while loops.

Superficially, I think, this code could be re-written to avoid a while loop.

Steps: instead of using the random.choice, just use `random.shuffle` and get a  shuffled list of  route_vectors and simply assign these route vectors to the first 4 (or max. no, of entry gates)  agents.  Thereafter, assign "not yet on the road" in a new "for" loop.
- Also, this way, you wouldn't have to check for "is_cell_vacant"

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -73,13 +116,96 @@ def __init__(self, grid_shape=(14, 14), n_agents=10, n_max=10, arrive_prob=0.5,
+            return True
+        return False
+
     def get_agent_obs(self):
```
In general, this method has a lot of "for" loops that could be merged

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
 
         return _grid
 
+    def step(self, agents_action):
```
In general, this method has a lot of "for" loops that can be merged.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
+            " dimension but was found to be of {}".format(self.n_agents, len(agents_action))
+
+        # Todo: Add check to validate each sub-action. It should be a valid action.
+        # This must be done before position of any car is updated
```
Add a check to validate each sub-action before updating any of the agent's position.

As of now, it is being checked in `__update_agent_pos`, but that checking it over there would lead to position updates to some agents.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
+
+        if self._step_count >= self._max_steps:
+            for i in range(self.n_agents):
+                self._agent_dones[i] = True
```
This for loop could be merged with  with "line 319"

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
+                self._agent_dones[i] = True
+
+        # gives additional step punishment to avoid jams
+        for agent_i in range(self.n_agents):
```
reward information is spread across 2 methods i.e over here(line 330) and  inside __update_agent_pos ( collision_reward).

Let's keep it in one place only. 

`Option 1`: move the step_cost inside update_agent_pos. 
`Option 2`: return collision as a "bool" from __update_agent_pos and then based on collision flag add the collision_reward.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +271,171 @@ def __create_grid(self):
+        # adds new car according to the probability _arrive_prob
+        if random.uniform(0, 1) < self._arrive_prob and self.curr_cars_count < self._n_max:
+            for agent_i in range(self.n_agents):
+                if self.__is_gate_free() and not self._on_the_road[agent_i]:
```
Instead of traversing through n_agents at every time step, we can do the following:

1. Maintain a  list of unassigned cars: Initially, all cars are in the unassigned list (stack) and we simply pop a car , every time it's added to the entry gate. And, every time, it finishes its route, we push it back to the list.
2. Get a List of Free gates and choose a free gate randomly.
3.  Pop a car from the unassigned list and put it on the gate.

koulanurag commented:
> Thanks, that is clear to me now. I have addressed the comments you left here. If you feel anything else needs to be changed please let me know.
> 
> Thanks again for your feedback.

Hi @rafaelmp2, Thanks for making the changes. We need to make some more changes before we wrap it up. Let me know if you would make the new suggested changes.

rafaelmp2 commented:
Hi @koulanurag thanks for the reviews. I made the changes you requested in the comments. Regarding your suggestion in the part to add more agents, I changed it in a slightly different way to avoid creating more data structures. Please check if everything is better now and let me know.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +270,175 @@ def __create_grid(self):
+
+
+        # verifies if any destination place was reached
+        for agent_i in range(self.n_agents):
```
Can this for-loop be merged with the for-loop on line #300 ?

rafaelmp2 replied:
I think we can merge them but an agent that arrives to a gate will not receive the step penalty for the first arriving time step

koulanurag commented:
Hi @rafaelmp2, Let me know if you can resolve the above `comment`.

Also, while you are at it, please add test cases for this environment. You can refer to https://github.com/koulanurag/ma-gym#testing and https://github.com/koulanurag/ma-gym/tree/master/tests/envs

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -88,9 +213,29 @@ def get_agent_obs(self):
+                        # get relative position for other agents location
+                        _other_agents_pos[row - (pos[0] - 1), col - (pos[1] - 1)] = 1
+
+            # set center value to 0 that belongs to agent_i TODO: see if this stays or not
```
Is this `Todo` done?

koulanurag replied:
While you are at it, can you also clean up other `Todo` which are done?

rafaelmp2 replied:
Sure will clean that

rafaelmp2 commented:
@koulanurag I've pushed the new version. The only `todo` i left was the one you added yourself in the introductory comments. I will add the tests asap too.

koulanurag commented:
Hi @rafaelmp2, I have done some minor refactoring. 

Also, superficially, the code looks good. I will try to do a thorough re-review in 1-2 days.  In the meanwhile, I assume you would be adding test cases.

`Optional`: It would be nice if the environment could be made more visually appealing. One thing that can be done is to use `white color` only for the 4 big walls in each quadrant. Another interesting aspect could be `emojis` for creating the visual environments such as using their cars and greenery as a fill-up for the image. Though, I am not sure if this will make it look better.

rafaelmp2 commented:
This was just to fix the observation space to make it work with the test files. I will pushed them soon.

koulanurag commented on tests/envs/test_tafficjunction.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def test_init(env):
+    assert 1 <= env._n_max <= 10, f'N_max must be between 1 and 10, got {env._n_max}'
```
`{env._n_max}` is not compatible with `python 3.5`. 

Let's just use:
```
assert 1 <= env._n_max <= 10, 'N_max must be between 1 and 10, got {}'.format(env._n_max)
```

Also, the issue is across the entire test file.

koulanurag replied:
Todo for me: I will add "3.5" constraints in the workflow.

koulanurag commented:
Hi @rafaelmp2, Thanks for the work so far.

Let me know once you are done with all the changes, and, I can go ahead and do a review.

rafaelmp2 commented:
Thanks @koulanurag . I think the only missing change now is your suggestion to improve the visual appearance of the environment. I will try to improve that and will let you know once I am done. Regarding the rest, at a first glance I think I have made all the changes.

koulanurag commented:
Just some clean-up and python 3.5 compatibility.

rafaelmp2 commented:
Hi @koulanurag , I changed the colors, it looks better now in my opinion. I don't understand exactly what you mean with the emojis idea and since i didn't want to change the draw file in the utils folder I just changed the colors as you suggested. Let me know what you think. I think after this you can go ahead with your review as you said.

koulanurag commented:
Hi @rafaelmp2, The visual changes looked good!

Also, I guess, there was some misunderstanding in the observation for each agent. I guess it's correctly mentioned now in the description of the env.

 And, there is an `issue` in reward for each agent. I have highlighted that as `todo` in the code.

If you can resolve it, that would be great. It will take me more than a day to get back to it,

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +282,173 @@ def __create_grid(self):
+                self._on_the_road[agent_i] = False
+                self.curr_cars_count -= 1
+
+            # Todo: Once a car reaches it's destination , it will never enter again in any of the tracks.
```
Only new cars enter the grid and all those cars which have reached their destination wouldn't be considered for entering any track.

Also, if all cars have reached their destination, then the episode terminates.

koulanurag commented on ma_gym/envs/traffic_junction/traffic_junction.py:
```diff
@@ -125,25 +282,173 @@ def __create_grid(self):
+
+            # gives additional step punishment to avoid jams
+            if self._on_the_road[agent_i]:
+                # Todo: To discourage a traffic jam, each car gets reward of `τ * r_time = −0.01τ`
```
We need to keep track of step_count for each agent since it entered the track and this step count will be used for penalty rather than total step count.

rafaelmp2 commented:
Thanks @koulanurag for letting me know about this. You are right, it was indeed my misunderstanding. I will address this asap.

rafaelmp2 commented:
I think it is good now, I've fixed the step cost for each agent and the episode termination/cars removal. 
Just a side note, `(n_agents, (3^2) × |n| × |l| × |r|)` is this the shape of the returned list of observations? Isn't it `(n_agents, (3^2) × (|n| + |l| + |r|))` instead as you wrote in other comments?

koulanurag commented on tests/envs/test_trafficjunction.py:
```diff
@@ -0,0 +1,151 @@
+        step_i = 0
+        done = [False for _ in range(env.n_agents)]
+        agent_0_route = obs[0].reshape((9, 9))[4][6:]  # one-hot
+        # Todo: Find max. steps for agent 0 based on it's route
```
Hi @rafaelmp2, Thanks for the work so far. We are almost done and just need to add some test cases involving rollouts for sanity.

In this particular test case, we need to ensure that once the agent reaches the destination; it will get "0" reward for rest of the episode.

I am not sure of the "minimal number of steps" to reach the destination required for a agent when there is no collision and agent has applied "gas" at all time-steps. This number varies based on the route taken by the agent. 
It will be helpful if you can finish this test-case by adding the current number of steps. Currently, it has wrong values,.

koulanurag replied:
It will also help if you can add more rollout test cases, such as a scenario where some number of agents collide so that we can ensure that it gets the correct collision reward.

koulanurag commented on tests/envs/test_trafficjunction.py:
```diff
@@ -0,0 +1,151 @@
+    obs = env.reset()
+    expected_agent_i_shape = (np.prod(env._agent_view_mask) * (env.n_agents + 2 + 3),)
+    for agent_i in range(env.n_agents):
+        assert obs[agent_i].shape == expected_agent_i_shape, \
```
I have added a test case to validate the observation shape.  For a single agent, it should be "3^2 * (|n|+|r|+|l|)"

koulanurag commented on ma_gym/__init__.py:
```diff
@@ -40,11 +40,12 @@
-    )
+    for num_max_cars in [4, 10]:
+        register(
+            id='TrafficJunction{}-v'.format(num_max_cars) + str(i),
```
Also, the name of env. has changed now, It will be either called TrafficJunction4-v0 or TrafficJunction10-v0 based on the number of maximum cars allowed.

koulanurag commented:
Here is a video of a trained agent.
![video 0_539_a156723eaf15ba7488b4](https://user-images.githubusercontent.com/4881224/124710282-66c9a200-deb1-11eb-94e5-cd8bf6e3f406.gif)

koulanurag commented:
Also, once this is done, We just need to write-up wiki for this environment which will be a copy-paste from the env. description in the code.

Current Wiki: [https://github.com/koulanurag/ma-gym/wiki/Environments#TrafficJunction](https://github.com/koulanurag/ma-gym/wiki/Environments#TrafficJunction)

rafaelmp2 commented:
Hi @koulanurag, thanks for adding those tests and recording the video. I fixed the max steps for the no collision test and added a case test where all agents collide in the middle of the junction. Regarding the wiki, I dont think I can edit, if you prefer I can copy it to my fork and you can copy it after. Let me know what you think.

koulanurag commented:
You can put the wiki content over here in the comments and I can put the content back in "Wiki"

koulanurag commented:
Hi @rafaelmp2, Thanks for all the work! We are finally done!

I will go ahead and merge it with the master repo.

rafaelmp2 commented:
Thanks @koulanurag! I will paste here the wiki soon.

rafaelmp2 commented:
This consists of a 4-way junction on a 14 × 14 grid. At each time step, "new" cars enter the grid with probability `p_arrive` from each of the four directions. However, the total number of cars at any given time is limited to `Nmax`. 
Each car occupies a single cell at any given time and is randomly assigned to one of three possible routes (keeping to the right-hand side of the road). At every time step, a car has two possible actions: gas which advances it by one cell on its route or brake to stay at its current location. A car will be removed once it reaches its destination at the edge of the grid.

Two cars collide if their locations overlap. A collision incurs a reward `rcoll = −10`, but does not affect the simulation in any other way. To discourage a traffic jam, each car gets reward of `τ * r_time = −0.01τ` at every time step, where `τ` is the number time steps passed since the car arrived. Therefore, the total reward at time t is 
<p align="center">
<img src="https://latex.codecogs.com/gif.latex?r(t)&space;=&space;C^t&space;*&space;r_{coll}&space;&plus;&space;\sum_{i=1}^{N^t}&space;{\tau_i&space;*&space;r_{time}}" height="50"/> 
</p>
where C^t is the number of collisions occurring at time t and N^t is number of cars present. The simulation is terminated after 'max_steps(default:40)' steps and is classified as a failure if one or more collisions have occurred.

Each car is represented by one-hot binary vector set {n, l, r}, that encodes its unique ID, current location and assigned route number respectively. Each agent controlling a car can only observe other cars in its vision range (a surrounding 3 × 3 neighborhood), though low level communication is allowed in "v1" version of the game. 

The state vector s_j for each agent is thus a concatenation of all these vectors, having dimension (3^2) × (|n| + |l| + |r|).

Reference : Learning Multi-agent Communication with Backpropagation
Url : https://papers.nips.cc/paper/6398-learning-multiagent-communication-with-backpropagation.pdf

`Versions`
| Name | Description |
| -------- | --------------- |
|TrafficJunction4-v0|Each agent get's it's own local observation|
|TrafficJunction4-v1|Each agent get's local observation of every other agent|

Same variants are available for 10 agents as `TrafficJunction10-v{0, 1}`

<p align="center">
<img src="https://user-images.githubusercontent.com/4881224/124710282-66c9a200-deb1-11eb-94e5-cd8bf6e3f406.gif" height="300"</img>
</p>

rafaelmp2 commented:
@koulanurag just realized, in case you cant access the raw text of this, I will just leave here the equation bit:
```
<p align="center">
<img src="https://latex.codecogs.com/gif.latex?r(t)&space;=&space;C^t&space;*&space;r_{coll}&space;&plus;&space;\sum_{i=1}^{N^t}&space;{\tau_i&space;*&space;r_{time}}" height="50"/> 
</p>
```

koulanurag commented:
I have added it.

[https://github.com/koulanurag/ma-gym/wiki/Environments#TrafficJunction](https://github.com/koulanurag/ma-gym/wiki/Environments#TrafficJunction)
